### PR TITLE
ACM-13449: Label ConfigMaps and Secrets for backup

### PR DESCRIPTION
**Pull request by carbonin**
In order for the ACM hub DR solution to include our required ConfigMaps and Secrets, they must be labeled with
`cluster.open-cluster-management.io/backup=true`.

Ensure the label is present on each reconcile.

The ImageClusterInstall is already covered because all resources in the `hive.openshift.io` group are automatically backed up.

Resolves https://issues.redhat.com/browse/ACM-13449

### CODE REVIEW

File: controllers/imageclusterinstall_controller.go [hunk 1]
```diff
@@ -574,6 +578,89 @@ func (r *ImageClusterInstallReconciler) removeBMHImage(ctx context.Context, bmhR
...
+	obj.SetLabels(labels)
+	return true
+}
+
+func (r *ImageClusterInstallReconciler) labelConfigMapForBackup(ctx context.Context, key types.NamespacedName) error {
+	cm := &corev1.ConfigMap{}
+	if err := r.Get(ctx, key, cm); err != nil {
+		return err
+	}
+
```
Comment (eranco74) [inline controllers/imageclusterinstall_controller.go:600]: Consider checking if the label already exists and return

File: controllers/imageclusterinstall_controller.go [hunk 2]
```diff
@@ -574,6 +578,89 @@ func (r *ImageClusterInstallReconciler) removeBMHImage(ctx context.Context, bmhR
...
+	}
+	return nil
+}
+
+func (r *ImageClusterInstallReconciler) labelSecretForBackup(ctx context.Context, key types.NamespacedName) error {
+	secret := &corev1.Secret{}
+	if err := r.Get(ctx, key, secret); err != nil {
+		return err
+	}
+
```
Comment (eranco74) [inline controllers/imageclusterinstall_controller.go:613]: Consider checking if the label already exists and return

File: controllers/imageclusterinstall_controller.go [hunk 3]
```diff
@@ -574,6 +578,89 @@ func (r *ImageClusterInstallReconciler) removeBMHImage(ctx context.Context, bmhR
...
+			log.WithError(err).Errorf("failed to label ConfigMap %s for backup", manifestKey)
+		}
+	}
+
+	if cd.Spec.PullSecretRef != nil {
+		psKey := types.NamespacedName{Name: cd.Spec.PullSecretRef.Name, Namespace: cd.Namespace}
+		if err := r.labelSecretForBackup(ctx, psKey); err != nil {
+			log.WithError(err).Errorf("failed to label Secret %s for backup", psKey)
+		}
+	}
```
Comment (eranco74) [inline controllers/imageclusterinstall_controller.go:661]: What about ImageSetRef and ClusterDeploymentRef?

File: controllers/imageclusterinstall_controller.go (hunk 1 above)
Comment (carbonin) [reply controllers/imageclusterinstall_controller.go:600]: This is being done based on the return value of `setBackupLabel`.

I did it this way to share the logic between secrets and config maps.

File: controllers/imageclusterinstall_controller.go (hunk 2 above)
Comment (carbonin) [reply controllers/imageclusterinstall_controller.go:613]: Same response as above.

File: controllers/imageclusterinstall_controller.go (hunk 3 above)
Comment (carbonin) [reply controllers/imageclusterinstall_controller.go:661]: ClusterImageSets and ClusterDeployments are included in the backup by default.

**Comment by carbonin** [comment]
Just FYI see https://docs.redhat.com/en/documentation/red_hat_advanced_cluster_management_for_kubernetes/2.10/html/business_continuity/business-cont-overview#resources-that-are-backed-up for the list of objects being backed up by default

**Review by eranco74 — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**